src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -235,6 +238,61 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     }
   }
 
+  private def verifyNoSubmissions(userRef: RawlsUserRef, dataAccess: DataAccess): ReadAction[PerRequestMessage] = {
+    dataAccess.submissionQuery.findBySubmitter(userRef.userSubjectId.value).exists.result flatMap {
+      case true => DBIO.failed(new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Cannot delete a user with submissions")))
+      case _ => DBIO.successful(RequestComplete(StatusCodes.OK))
+    }
+  }
+
+  // can't just delete DB rows; also removes user from Google Groups
+  private def removeUserFromAllGroups(userRef: RawlsUserRef, dataAccess: DataAccess): ReadWriteAction[RawlsUser] = {
+    withUser(userRef, dataAccess) { user =>
+      dataAccess.rawlsGroupQuery.listGroupsForUser(userRef) flatMap { groupRefs =>
+        // Slick complains if it remains a Set
+        val removeActions = groupRefs.toSeq.map { groupRef =>
+          withGroup(groupRef, dataAccess) { group =>
+            updateGroupMembersInternal(group, Set(user), Set.empty, RemoveGroupMembersOp, dataAccess)
```

Comment:
... and my solution is actually wrong because `listGroupsForUser` recurses through group membership, and therefore it tries to remove the user at every level.  So I'll fix that too.